### PR TITLE
Implementation of Sink extension

### PR DESCRIPTION
**Pull request by CharukaK**
## Purpose
This will add Sink implementation to the extension.

## User stories
[https://docs.google.com/document/d/168kc6rESZvsUWecoW01AomdS6B-JE0d9ks4ADei2-L0/edit?usp=sharing](https://docs.google.com/document/d/168kc6rESZvsUWecoW01AomdS6B-JE0d9ks4ADei2-L0/edit?usp=sharing)

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Test environment
JDK v1.8.0_161
 
## Learning
https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/Welcome.html

### CODE REVIEW

File: component/src/main/java/org/wso2/extension/siddhi/io/sqs/sink/SQSMessagePublisher.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        SendMessageRequest sendMessageRequest = new SendMessageRequest()
+                .withQueueUrl(sqsSinkConfig.getQueueUrl())
+                .withMessageBody(message);
+
+        if (sqsSinkConfig.getDelayIntervalTime() != -1) {
+            sendMessageRequest.withDelaySeconds(sqsSinkConfig.getDelayIntervalTime());
+        }
+
+        return sendMessageRequest;
+
```
Comment (minudika) [inline component/src/main/java/org/wso2/extension/siddhi/io/sqs/sink/SQSMessagePublisher.java]: remove unnecessary newlines

File: component/src/main/java/org/wso2/extension/siddhi/io/sqs/sink/SQSMessagePublisher.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                                        .getValue(dynamicOptions));
+            }
+
+        }
+
+        SendMessageResult sendMessageResult = amazonSQS.sendMessage(sendMessageRequest);
+
+        if (sendMessageResult.getSdkHttpMetadata().getHttpStatusCode() != SQSConstants.HTTP_SUCCESS) {
+            throw new ConnectionUnavailableException("Error occurred when trying to send the message, received http" +
+                    " status code : " + sendMessageResult.getSdkHttpMetadata().getHttpStatusCode());
```
Comment (minudika) [inline component/src/main/java/org/wso2/extension/siddhi/io/sqs/sink/SQSMessagePublisher.java]: can't we get the cause of the failure from the result?

File: component/src/main/java/org/wso2/extension/siddhi/io/sqs/sink/SQSSink.java [hunk 3]
```diff
@@ -32,51 +35,90 @@
...
         name = "sqs",
         namespace = "sink",
-        description = " ",
+        description = "SQS sink allows users to connect and publish messages to an AWS SQS Queue. It has the" +
+                " ability to only publish Text messages",
         parameters = {
-
+                @Parameter(
+                        name = SQSConstants.QUEUE_URL_NAME,
+                        description = "Queue name which SQS Sink should subscribe to",
```
Comment (minudika) [inline component/src/main/java/org/wso2/extension/siddhi/io/sqs/sink/SQSSink.java]: this should be url of the queue

File: component/src/main/java/org/wso2/extension/siddhi/io/sqs/sink/SQSSink.java [hunk 4]
```diff
@@ -116,7 +158,17 @@
...
+        this.sqsMessagePublisher = null;
+
+        if (this.sinkConfig.getAccessKey() == null) {
+            this.sinkConfig.setAccessKey(configReader.readConfig(SQSConstants.ACCESS_KEY_NAME, null));
+        }
 
+        if (this.sinkConfig.getSecretKey() == null) {
+            this.sinkConfig.setSecretKey(configReader.readConfig(SQSConstants.SECRET_KEY_NAME, null));
+        }
     }
```
Comment (minudika) [inline component/src/main/java/org/wso2/extension/siddhi/io/sqs/sink/SQSSink.java]: what if one/both of above "accessKey" and "secretKey" are not provided?

File: component/src/main/java/org/wso2/extension/siddhi/io/sqs/util/SQSConstants.java [hunk 5]
```diff
@@ -23,11 +23,13 @@
...
-    public static final String QUEUE_NAME = "queue";
+    // general properties for sqs client
     public static final String ACCESS_KEY_NAME = "access.key";
     public static final String SECRET_KEY_NAME = "secret.key";
     public static final String REGION_NAME = "region";
+    public static final String QUEUE_URL_NAME = "queue";
+
+    // Option parameter names for source
     public static final String POLLING_INTERVAL_NAME = "polling.interval";
     public static final String WAITING_TIME_NAME = "waiting.time";
```
Comment (minudika) [inline component/src/main/java/org/wso2/extension/siddhi/io/sqs/util/SQSConstants.java]: use terms used in Amazon SQS

File: component/src/test/java/org/wso2/extension/siddhi/io/sqs/sink/TestCaseOfSQSSink.java [hunk 6]
```diff
@@ -1,7 +1,51 @@
...
+import org.testng.annotations.Test;
+import org.wso2.siddhi.core.SiddhiAppRuntime;
+import org.wso2.siddhi.core.SiddhiManager;
+import org.wso2.siddhi.core.stream.input.InputHandler;
+
+import java.util.Date;
+
 public class TestCaseOfSQSSink {
     // If you will know about this related testcase,
     //refer https://github.com/wso2-extensions/siddhi-io-file/blob/master/component/src/test
```
Comment (minudika) [inline component/src/test/java/org/wso2/extension/siddhi/io/sqs/sink/TestCaseOfSQSSink.java]: Remove  unnecessary comments.. siddhi-io-file?

File: component/src/test/java/org/wso2/extension/siddhi/io/sqs/source/TestCaseForSourceWithSiddhiAppRuntime.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+import org.wso2.siddhi.core.query.output.callback.QueryCallback;
+import org.wso2.siddhi.core.util.EventPrinter;
+
+public class TestCaseForSourceWithSiddhiAppRuntime {
+
+    /**
+     * For this Test case Follow the following instructions.
+     * 1) Create any type of queue in a Service
+     * 2) Replace the queue url, access key, secret key in proper places with the ones relevant to the queue.
+     * 3) Uncomment and run the Test.
```
Comment (minudika) [inline component/src/test/java/org/wso2/extension/siddhi/io/sqs/source/TestCaseForSourceWithSiddhiAppRuntime.java]: comment these tests in testng.xml and add relevant instructions here.

File: component/src/main/java/org/wso2/extension/siddhi/io/sqs/sink/SQSMessagePublisher.java (hunk 2 above)
Comment (CharukaK) [reply component/src/main/java/org/wso2/extension/siddhi/io/sqs/sink/SQSMessagePublisher.java]: SendMessageResult object will only provide the status code of the response.

https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/CommonErrors.html

File: component/src/main/java/org/wso2/extension/siddhi/io/sqs/source/SQSSource.java [hunk 8]
```diff
@@ -186,7 +207,7 @@ public void init(SourceEventListener sourceEventListener, OptionHolder optionHol
      */
     @Override
     public Class[] getOutputEventClasses() {
-        return new Class[]{String.class}; // SQS message body supports only text.
+        return new Class[] {String.class}; // SQS message body supports only text.
```
[the remark this answers is not in the document]
Comment (CharukaK) [reply component/src/main/java/org/wso2/extension/siddhi/io/sqs/source/SQSSource.java]: The message body only supports Text messages.

**Pull request merged**